### ISSUE TITLE
Default verify_ap_req_nofail used by krb5_verify_init_creds is insecure

### ISSUE THREAD

**Issue by riastradh** [heimdal/heimdal#1129]
**Describe the bug**
The default value of [libdefaults] verify_ap_req_nofail is false.

This means that applications which rely on krb5_verify_init_creds, such as pam_krb5, are vulnerable to KDC spoofing attacks by default.

I suspect this option was created for legacy _mit-krb5_ applications which were (foolishly) made to rely on krb5_verify_init_creds without a keytab to authenticate the KDC, because in mit-krb5, krb5_verify_init_creds is allowed to succeed when there is no keytab:

> If the specified keytab does not exist, or is empty, or cannot be read, or does not contain an entry for server , then credential verification may be skipped unless configuration demands that it succeed. The caller can control this behavior by providing a verification options structure; see [krb5_verify_init_creds_opt_init()](https://web.mit.edu/kerberos/krb5-1.20/doc/appdev/refs/api/krb5_verify_init_creds_opt_init.html#c.krb5_verify_init_creds_opt_init) and [krb5_verify_init_creds_opt_set_ap_req_nofail()](https://web.mit.edu/kerberos/krb5-1.20/doc/appdev/refs/api/krb5_verify_init_creds_opt_set_ap_req_nofail.html#c.krb5_verify_init_creds_opt_set_ap_req_nofail).
>
> https://web.mit.edu/kerberos/krb5-1.20/doc/appdev/refs/api/krb5_verify_init_creds.html

However, in _Heimdal_, krb5_verify_init_creds has failed immediately when there is no keytab as far back as 1997 when krb5_verify_init_creds was first filled out:

https://github.com/heimdal/heimdal/blob/d5b74e06b0e655083367e69d53b2394bbb74cee6/lib/krb5/verify_init.c#L104-L109

So it appears to me this is only really relevant for legacy realms where there _is_ a keytab that the application _could_ use to authenticate the KDC, and the default-off option is just waiting to shoot you in the foot.

I suggest turning this option on by default so that a default (empty) /etc/krb5.conf is safe by default.  Of course, legacy applications that rely on allowing a fraudulent KDC to function can still turn it off.

**To Reproduce**
code inspection and history spelunking

**Expected behavior**
krb5_verify_init_creds authenticates the KDC with its keytab, and treats spoofed KDC as authentication failure.

**Comment by nicowilliams** [comment]
> The default value of [libdefaults] verify_ap_req_nofail is false.

Correct, which means that the default is _to fail_ if there is no service principal and key to verify the user's TGT with.  That is the correct and secure default.

That said, we don't have a _test_ of this.
